src/app/file-panel.js

```diff
@@ -108,14 +112,16 @@ function filepanel (appAPI, files) {
 
   var events = new EventManager()
   var element = template()
-  // TODO please do not add custom javascript objects, which have no
-  // relation to the DOM to DOM nodes
-  element.events = events
+
   fileExplorer.events.register('focus', function (path) {
     appAPI.switchToFile(path)
   })
 
-  return element
+  var api = { events: events }
+
+  container.appendChild(element)
+
+  return api
```

Comment:
i think version **Number 4** is the most traditional version that existed already in ECMA Script 3.

---

#### Usage (my personal recommendation without `this`)
```js
var foobarAPI = {}
var foobarEvents = new EventManager()
var opts = { }
var foobar = FooBar('.container', foobarAPI, foobarEvents, opts)
// but
// var foobar = new FooBar('.container', foobarAPI, foobarEvents, opts) // works exactly the same
```

##### Definition  (my personal recommendation would be something like)
```js
var css = csjs`.foobar { font-size: 20px; }`
function FooBar (container, appAPI, appEvents, opts) {
  if (typeof container === 'string') container = document.querySelector(container)
  if (!container) throw new Error('no container given')
  /* ... */
  var el = bel`<h1 class=${css.foobar}> hello ${opts.name || 'world'} </h1>`
  container.appendChild(el)
  /* ... */
  var api = new EventEmitter()
  api.name = opts.name || 'anoymous'
  api.__proto__ = methods
  return api
}
var methods = {
  bar: function () {},
  baz: function () {}
}
```